File: coolero/services/device_commander.py
#  Coolero - monitor and control your cooling and other devices
#  Copyright (c) 2021  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging
from typing import Callable

from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.date import DateTrigger

from coolero.models.device import DeviceType
from coolero.models.lighting_mode import LightingModeType
from coolero.models.settings import Setting
from coolero.models.speed_profile import SpeedProfile
from coolero.repositories.hwmon_repo import HwmonRepo
from coolero.repositories.liquidctl_repo import LiquidctlRepo
from coolero.services.dynamic_controls.lighting_controls import LightingControls
from coolero.services.notifications import Notifications
from coolero.services.speed_scheduler import SpeedScheduler
from coolero.services.utils import MathUtils
from coolero.settings import Settings as SavedSettings
from coolero.view.uis.canvases.speed_control_canvas import SpeedControlCanvas

_LOG = logging.getLogger(__name__)


class DeviceCommander:

    def __init__(self,
                 lc_repo: LiquidctlRepo,
                 hwmon_repo: HwmonRepo,
                 base_scheduler: BackgroundScheduler,
                 speed_scheduler: SpeedScheduler,
                 notifications: Notifications) -> None:
        self._lc_repo = lc_repo
        self._hwmon_repo: HwmonRepo = hwmon_repo
        self._base_scheduler = base_scheduler
        self._job_id: str = 'device_setting'
        self._speed_scheduler: SpeedScheduler = speed_scheduler
        self._notifications: Notifications = notifications

    def set_speed(self, subject: SpeedControlCanvas) -> None:
        channel: str = subject.channel_name
        device_id: int = subject.device.type_id
        if subject.current_speed_profile == SpeedProfile.FIXED:
            setting = Setting(channel, speed_fixed=subject.fixed_duty, pwm_mode=subject.pwm_mode)
            SavedSettings.save_fixed_profile(
                subject.device.name, device_id, channel, subject.current_temp_source.name, subject.fixed_duty,
                subject.pwm_mode
            )
            SavedSettings.save_applied_fixed_profile(
                subject.device.name, device_id, channel, subject.current_temp_source.name, subject.fixed_duty,
                subject.pwm_mode
            )
        elif subject.current_speed_profile == SpeedProfile.CUSTOM:
            setting = Setting(
                channel,
                speed_profile=MathUtils.convert_axis_to_profile(subject.profile_temps, subject.profile_duties),
                temp_source=subject.current_temp_source,
                pwm_mode=subject.pwm_mode
            )
            SavedSettings.save_custom_profile(
                subject.device.name, device_id, channel, subject.current_temp_source.name,
                subject.profile_temps, subject.profile_duties, subject.pwm_mode
            )
            SavedSettings.save_applied_custom_profile(
                subject.device.name, device_id, channel, subject.current_temp_source.name,
                subject.profile_temps, subject.profile_duties, subject.pwm_mode
            )
        elif subject.current_speed_profile in [SpeedProfile.NONE, SpeedProfile.DEFAULT]:
            SavedSettings.save_applied_none_default_profile(
                subject.device.name, device_id, channel, subject.current_temp_source.name,
                subject.current_speed_profile, subject.pwm_mode
            )
            scheduled_setting_removed: bool = self._speed_scheduler.clear_channel_setting(subject.device, channel)
            setting = Setting(channel, pwm_mode=subject.pwm_mode)
            if subject.device.type == DeviceType.HWMON and self._hwmon_repo is not None:
                _LOG.info('Scheduling settings for %s', subject.device.name)
                _LOG.debug('Scheduling speed device settings: %s', setting)
                self._add_to_device_jobs(
                    lambda: self._notifications.settings_applied(
                        self._hwmon_repo.set_channel_to_default(device_id, setting)
                    )
                )
            elif subject.device.type == DeviceType.LIQUIDCTL and scheduled_setting_removed:
                _LOG.info('Cleared scheduled device setting for %s', subject.device.name)
                self._add_to_device_jobs(
                    lambda: self._notifications.settings_applied(f'{subject.device.name} - removed')
                )
            return
        else:
            setting = Setting('none')
        _LOG.info('Scheduling settings for %s', subject.device.name)
        _LOG.debug('Scheduling speed device settings: %s', setting)
        self._speed_scheduler.clear_channel_setting(subject.device, channel)
        # Requirements to use our internal scheduler:
        if subject.current_speed_profile == SpeedProfile.CUSTOM \
                and ((subject.device != subject.current_temp_source.device
                      and subject.current_temp_source.device.info.temp_ext_available)
                     or (subject.device == subject.current_temp_source.device
                         and subject.device.info.channels[channel].speed_options.manual_profiles_enabled)):
            self._notifications.settings_applied(
                self._speed_scheduler.set_settings(subject.device, setting)
            )
        # liquidctl devices not meeting the above criteria can handle profiles themselves, and fixed speeds:
        elif subject.device.type == DeviceType.LIQUIDCTL:
            self._add_to_device_jobs(
                lambda: self._notifications.settings_applied(
                    self._lc_repo.set_settings(device_id, setting)
                )
            )
        # hwmon fixed speeds
        elif subject.device.type == DeviceType.HWMON:
            self._add_to_device_jobs(
                lambda: self._notifications.settings_applied(
                    self._hwmon_repo.set_settings(device_id, setting)
                )
            )

    def set_lighting(self, subject: LightingControls) -> None:
        if subject.current_set_settings is None:
            return
        device_id, lighting_setting = subject.current_set_settings
        SavedSettings.save_lighting_settings()
        if lighting_setting.lighting_mode.type != LightingModeType.LC:
            return  # only LC lighting modes are currently supported
        _LOG.info('Scheduling lighting settings for Liquidctl device #%s', device_id)
        _LOG.debug('Scheduling lighting device settings: %s', lighting_setting)
        self._add_to_device_jobs(
            lambda: self._notifications.settings_applied(
                self._lc_repo.set_settings(device_id, lighting_setting)
            )
        )

    def reinitialize_devices(self) -> None:
        self._add_to_device_jobs(
            self._lc_repo.reinitialize_devices
        )

    def _add_to_device_jobs(self, set_function: Callable) -> None:
        self._base_scheduler.add_job(
            set_function,
            DateTrigger(),  # defaults to now()
            id=self._job_id
        )

File: coolercontrold/src/device_commander.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2022  Guy Boldon
 * |
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * |
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * |
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 ******************************************************************************/

use std::collections::HashMap;
use std::sync::Arc;

use anyhow::{anyhow, Context, Result};

use crate::{AllDevices, Repos};
use crate::config::Config;
use crate::device::DeviceType;
use crate::repositories::repository::Repository;
use crate::setting::Setting;
use crate::speed_scheduler::SpeedScheduler;

pub type ReposByType = HashMap<DeviceType, Arc<dyn Repository>>;

pub struct DeviceCommander {
    all_devices: AllDevices,
    repos: ReposByType,
    pub speed_scheduler: Arc<SpeedScheduler>,
}

impl DeviceCommander {
    pub fn new(all_devices: AllDevices, repos: Repos, config: Arc<Config>) -> Self {
        let mut repos_by_type = HashMap::new();
        for repo in repos.iter() {
            match repo.device_type() {
                DeviceType::CPU => repos_by_type.insert(DeviceType::CPU, Arc::clone(repo)),
                DeviceType::GPU => repos_by_type.insert(DeviceType::GPU, Arc::clone(repo)),
                DeviceType::Liquidctl => repos_by_type.insert(DeviceType::Liquidctl, Arc::clone(repo)),
                DeviceType::Hwmon => repos_by_type.insert(DeviceType::Hwmon, Arc::clone(repo)),
                DeviceType::Composite => repos_by_type.insert(DeviceType::Composite, Arc::clone(repo)),
            };
        }
        let speed_scheduler = Arc::new(SpeedScheduler::new(
            all_devices.clone(),
            repos_by_type.clone(),
            config,
        ));
        DeviceCommander { all_devices, repos: repos_by_type, speed_scheduler }
    }

    pub async fn set_setting(&self, device_uid: &String, setting: &Setting) -> Result<()> {
        if let Some(device_lock) = self.all_devices.get(device_uid) {
            let device_type = device_lock.read().await.d_type.clone();
            return if let Some(repo) = self.repos.get(&device_type) {
                if let Some(true) = setting.reset_to_default {
                    self.speed_scheduler.clear_channel_setting(device_uid, &setting.channel_name).await;
                    if device_type == DeviceType::Hwmon || device_type == DeviceType::GPU {
                        repo.apply_setting(device_uid, setting).await
                    } else {
                        Ok(()) // nothing to actually set in this case, just clear settings.
                    }
                } else if setting.speed_fixed.is_some() {
                    self.speed_scheduler.clear_channel_setting(device_uid, &setting.channel_name).await;
                    repo.apply_setting(device_uid, setting).await
                } else if setting.lighting.is_some() {
                    repo.apply_setting(device_uid, setting).await
                } else if setting.speed_profile.is_some() {
                    let speed_options = device_lock.read().await
                        .info.as_ref().with_context(|| "Looking for Device Info")?
                        .channels.get(&setting.channel_name).with_context(|| "Looking for Channel Info")?
                        .speed_options.clone().with_context(|| "Looking for Channel Speed Options")?;
                    if speed_options.profiles_enabled {
                        repo.apply_setting(device_uid, setting).await
                    } else if let None = setting.temp_source {
                        Err(anyhow!("A Temp Source must be set when scheduling a Speed Profile for this device: {}", device_uid))
                    } else if (
                        &setting.temp_source.as_ref().unwrap().device_uid == device_uid
                            && speed_options.manual_profiles_enabled)
                        || &setting.temp_source.as_ref().unwrap().device_uid != device_uid {
                        self.speed_scheduler.schedule_setting(device_uid, setting).await
                    } else {
                        Err(anyhow!("Speed Profiles not enabled for this device: {}", device_uid))
                    }
                } else if setting.lcd.is_some() {
                    let has_lcd_modes = !device_lock.read().await
                        .info.as_ref().with_context(|| "Looking for Device Info")?
                        .channels.get(&setting.channel_name).with_context(|| "Looking for Channel Info")?
                        .lcd_modes.is_empty();
                    if has_lcd_modes {
                        repo.apply_setting(device_uid, setting).await
                    } else {
                        Err(anyhow!("LCD Screen modes not enabled for this device: {}", device_uid))
                    }
                } else {
                    Err(anyhow!("Invalid Setting combination: {:?}", setting))
                }
            } else {
                Err(anyhow!("Repository: {:?} for device is currently not running!", device_type))
            };
        }
        {
            Err(anyhow!("Device Not Found: {}", device_uid))
        }
    }

    /// This is used to reinitialize liquidctl devices after waking from sleep
    pub async fn reinitialize_devices(&self) {
        if let Some(liquidctl_repo) = self.repos.get(&DeviceType::Liquidctl) {
            liquidctl_repo.reinitialize_devices().await;
        }
    }
}